### DISCUSSION TITLE
[How To] Install Linux Dependencies in Container

### DISCUSSION THREAD

**Discussion by LanceMcCarthy** [microsoft/aspire#4112] [Q&A] [answered]
Hi Folks!

Until now, I've always have been able to just install my required Linux dependencies at the top of my Dockerfile in the `base` section. For example:

![image](https://github.com/dotnet/aspire/assets/3520532/7f4a6944-37d9-4400-8f2a-a7fde28fa6c1)

However, I can't find a way to do this when building a .NET Aspire project 🙁

### Attempts

I tried finagling it by defining a command in the aspire-manifest, but it's not an option when using a "project.v0" type (I only see [expressing args when using the "docker.v0" type](https://learn.microsoft.com/en-us/dotnet/aspire/whats-new/preview-5?tabs=dotnet-cli#api-for-expressing-docker-build-args)).

![image](https://github.com/dotnet/aspire/assets/3520532/d757678f-9d8f-445e-864c-6d6939ea6ae3)

A better question might be "Can I run a bash command when a .NET Aspire container is first created?".

Thanks!
Lance

**Comment by timheuer** [comment]
I assume when running locally these are no issue as this is an ASP.NET project right? And you want to eliminate your previous dockerfile?

**Comment by LanceMcCarthy** [comment]
There is no Dockerfile right now, it's using the default `project.v0` approach that seems to use the donet SDK container build commands internally?

You can see everything here => https://github.com/LanceMcCarthy/DevCraftAspire

**Comment by LanceMcCarthy** [comment]
To explain a little further, my API project has a `Telerik.Drawing.Skia` NuGet package reference ([link to csproj](https://github.com/LanceMcCarthy/DevCraftAspire/blob/a742f28df6ca37ed393c6f7215cf5fd790363e0a/DevCraftAspire.ApiService/DevCraftAspire.ApiService.csproj#L11)).

That needs the following Linux dependencies to be available at runtime:

- `libfreetype6`
- `libfontconfig1`

I have no problem meeting this requirement when I'm deploying a standard ASP.NET project with a Dockerfile. [Here's a great real-world example](https://github.com/LanceMcCarthy/DevOpsExamples/blob/main/src/AspNetCore/MyAspNetCoreApp/Dockerfile) wher eI use Dockerfile to meet this requirement. Proving that it works, you can spin up the image using the Dockerhub 

`docker run -d  -p 8080:8080 lancemccarthy/myaspnetcoreapp:latest`

the problem arises when I am using Aspire to create the image. Where can I tell .NET to include those two dependencies in the image?

**Reply by timheuer** [reply]
I'm not able to get the project running locally to try things out.  I was hoping to see if `.WithArgs` on the web project would work in this regard...this would be translated to `args` in the manifest and I'm not sure how we pass those into the container image right now.  @davidfowl thoughts on this approach of adding image-specific dependencies on the resulting container image for a project-based resource easily (other than changing the base image to be an aspnet+yourdependencies image)?

**Reply by timheuer** [reply, marked as answer]
Ok, chatting with a few folks, you'd have to rely on a base image that has these dependencies already (assuming you don't want to use a dockerfile).  So you'd need an image that is your aspnet+dependencies added, then can set that as containerbaseimage in the web project and the rest of the workflow would work.

**Reply by LanceMcCarthy** [reply]
That makes sense and what I figured would be the JustWorks™ answer.

I'll just need to figure out the best workflow to make sure I get new base images generated whenever a new aspire image is released. Is there any guidance David and crew might be able to point me to automatically generate a new base image when aspire releases one?

I am [very familiar with publishing new multi-arch images to dockerhub in a workflow](https://github.com/LanceMcCarthy/DevOpsExamples/blob/57c74e9e7c2f75bfef2a8c5dfa5d0c9d25cf6631/.github/workflows/main_build-aspnetcore.yml#L18-L80), but not how to trigger a new build when an upstream repo has published something.

**Reply by timheuer** [reply]
I don't think your base image (which I interpret as aspnet:8.0 + your dependencies) needs to sync with aspire. All the Aspire packages (as of now) are part of your app pieces during deployment.  So you likely just want to sync an image of aspnet:<version> with your dependencies...much less rebuild.

**Reply by richlander** [reply]
The key point is that Lance is now on the hook to update this image separately as a new workflow.

**Reply by richlander** [reply]
This q is covered in the FAQ section: https://devblogs.microsoft.com/dotnet/streamline-container-build-dotnet-8/#frequently-asked-questions

**Reply by timheuer** [reply]
I mean you'd own it -- so setting `ContainerBaseImage` on the project that needs it.  again, I'm not sure it needs to be synced to any Aspire-related versions as I look at the projects.

**Reply by LanceMcCarthy** [reply]
Aha @richlander it's right there! I wish there was better SEO... but I guess that's what we are doing right now 😂

![image](https://github.com/dotnet/aspire/assets/3520532/bcea4a3a-8f41-4c37-b946-e538ebb69aad)

## Custom Base Images

Pretty much where I am right now... I've already built the custom base image easily enough [in this pre-build job ](https://github.com/LanceMcCarthy/DevCraftAspire/blob/1d17f2052f43feceeb40b2affbbc01ef4fc1492f/.github/workflows/build.yml#L9-L42 ) and here's the simple Dockerfile the job uses:

```Dockerfile
FROM mcr.microsoft.com/dotnet/aspnet:8.0 AS base

# Pre-install dependencies for Skia fonts
RUN apt-get update
RUN apt-get install libfreetype6 libfontconfig1 -y
```

It publishes to dockerhub as `lancemccarthy/aspirebase:8.0` (see https://hub.docker.com/r/lancemccarthy/aspirebase)
... but I was wondering how `azd deploy` was going to know how to use the new base image, because it doesnt have the same params as dotnet build.

If I understand what you're saying correct, I should be able to edit the csproj of the project with this (same as the CLI param)

```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <ContainerBaseImage>registry.hub.docker.com/lancemccarthy/aspirebase:8.0</ContainerBaseImage>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>
...
</Project>
```

Or maybe `lancemccarthy/aspirebase:8.0` instead? I'm about to YOLO it now... stay tuned https://github.com/LanceMcCarthy/DevCraftAspire/actions

**Reply by richlander** [reply]
>  I'm not sure it needs to be synced to any Aspire-related versions as I look at the projects.

This isn't the point. The "custom image" would need to be updated multiple times a month to having the same servicing characteristics as we have for the .NET base images.

Check out: https://github.com/mthalman/docker-bump-action

**Reply by LanceMcCarthy** [reply]
True, but you both answered my fundamental concerns:

- Tim clarified there was no custom "aspire" base image.
- Always push a new version of my `lancemccarthy/aspirebase:8.0` when `mcr.microsoft.com/dotnet/aspnet:8.0` is updated

I'm taking a look at `docker-bump-action` now. Seems like it only runs on a schedule and then updates the desired image when there's a diff with the base image. [Their example](https://github.com/mthalman/docker-bump-action-example/blob/main/.github/workflows/monitor-base-image-name.yml) looks very close to what I will end up doing to stay up to date.

Right now, it will always create a new base image before a new build gets pushed to Azure. This is because it's a required job before the Aspire app gets built 😉

```yaml
jobs:
  custom_base_image:
    ...

  deploy_app:
    needs: [custom_base_image]
```

### Problem

My only problem right now is trying to figure out why docker hub wants credentials when pulling a **public** image defined as my `ContainerBaseImage` in the csproj. [See the error in the azd deploy output.](https://github.com/LanceMcCarthy/DevCraftAspire/actions/runs/8992879895/job/24703652725)

```
Unable to access the repository 'lancemccarthy/aspirebase' in the registry 'docker.io'. Please confirm your credentials are correct and that you have access to this repository and registry. 
```

The `docker.io/lancemccarthy/aspirebase:8.0` image is public this shouldnt be happening. https://hub.docker.com/r/lancemccarthy/aspirebase 

It's also strange that it shows 12 pulls, could the error be due to something else the build is trying to do? While yes, I can log into docker with a single step, this is just not sound a security move... I don't want anything in that job pushing to the image repo.

**Reply by LanceMcCarthy** [reply]
Something is funky with setting the `ContainerBaseImage` setting, no matter the public source, it still fails with access error.

I switched to GitHub packages so the custom base image https://github.com/LanceMcCarthy/DevCraftAspire/pkgs/container/aspirebase (a public package), then I [changed ContainerBaseImage value in the cspro](https://github.com/LanceMcCarthy/DevCraftAspire/blob/main/DevCraftAspire.ApiService/DevCraftAspire.ApiService.csproj#L4)j to use it:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <ContainerBaseImage>ghcr.io/lancemccarthy/aspirebase:8.0</ContainerBaseImage>
    ...
  </PropertyGroup>
...
</Project>
```

but I get the same access problem when the project is compiled with `azd deploy`:

```
Unable to access the repository 'lancemccarthy/aspirebase' in the registry 'ghcr.io'. Please confirm your credentials are correct and that you have access to this repository and registry. [/home/runner/work/DevCraftAspire/DevCraftAspire/DevCraftAspire.ApiService/DevCraftAspire.ApiService.csproj]
```

### Review

- Here's the run, does anything stand out? https://github.com/LanceMcCarthy/DevCraftAspire/actions/runs/8993599394
- I think I followed all the rules from here https://github.com/dotnet/sdk-container-builds/blob/main/docs/ContainerCustomization.md#choosing-a-base-image

🤷🏼‍♂️

**Reply by timheuer** [reply]
@baronfel could be just a retry issue?

**Reply by baronfel** [reply]
@LanceMcCarthy have you logged in to ghcr.io via the Docker CLI before doing the `azd` deploy? Even public packages in ghcr require authentication - this took me by surprise too. There are detailed docs about how the SDK authenticates with registries [here](https://github.com/dotnet/sdk-container-builds/blob/main/docs/RegistryAuthentication.md), and I have a live example of this [in my testing repo](https://github.com/baronfel/sdk-container-demo/actions/runs/8647814886/workflow#L120-L125) if that helps demonstrate.

You don't strictly _need_ to use the Docker CLI to authenticate, but you do need a config json with the appropriate format.

**Reply by baronfel** [reply]
Separately we have a bug around Docker Hub truly-public images that currently also requires authentication - this enhancement is tracked [here](https://github.com/dotnet/sdk-container-builds/issues/406).

**Reply by LanceMcCarthy** [reply]
I already logged off for the day but I'll try it tomorrow morning.

I did try with docker cli login for docker hub, didn't work. However, I haven't tried it with the ghcr login yet.

Thanks!

**Reply by LanceMcCarthy** [reply]
Thank you @baronfel. Unfortunately, I'm still not able to authenticate with the registry. I've tried both `docker/login-action@v3` and `azure/docker-login@v1` that your demo uses.

## Setup & Errors

You can go right to the relevant items with the following links:

- [The csproj with the custom ContainerBaseImage defined](https://github.com/LanceMcCarthy/DevCraftAspire/blob/main/DevCraftAspire.ApiService/DevCraftAspire.ApiService.csproj#L4)
    - `<ContainerBaseImage>ghcr.io/lancemccarthy/aspirebase:8.0</ContainerBaseImage>`
- [The workflow steps](https://github.com/LanceMcCarthy/DevCraftAspire/actions/runs/9002706664/workflow)
    - ![image](https://github.com/dotnet/aspire/assets/3520532/9634979a-0286-4883-a19e-ed20b5624850) 
- [The failed workflow results](https://github.com/LanceMcCarthy/DevCraftAspire/actions/runs/9002706664/job/24731646817)
    - `error CONTAINER1016: Unable to access the repository 'lancemccarthy/aspirebase' in the registry 'ghcr.io'. Please confirm your credentials are correct and that you have access to this repository and registry. ` 

## Results

I can confirm the login is good:
![image](https://github.com/dotnet/aspire/assets/3520532/7259610a-0a20-45d2-8bda-31f4e1b921f2)

Is the `DOCKER_CONFIG` environment variable that the login step creates what you're ferring to when you say _"but you do need a config json with the appropriate format."_

**Reply by baronfel** [reply]
Yep, that should be all you have to do - can you confirm that after that login 'docker pull' can pull your image successfully?

**Reply by LanceMcCarthy** [reply]
@baronfel the pull succeeded:

![image](https://github.com/dotnet/aspire/assets/3520532/bd325fd8-e543-49b8-a370-cba8bd9f0b1a)


Do you think this might be happening because of `azd deploy`? If I use `dotnet build` command as a pre-step, how can I ensure azd deploy doesn't try to build again (I dont see a `--no-build` flag.

**Reply by baronfel** [reply]
> Do you think this might be happening because of azd deploy? If I use dotnet build command as a pre-step, how can I ensure azd deploy doesn't try to build again (I dont see a --no-build flag.

I don't have a clear reason to suspect yet - the next step would be to `dotnet publish /t:PublishContainer` on your specific project to see if the SDK itself can containerize the app. If the standalone command succeeds but the `azd deploy`-wrapped command fails, then we'd have clear evidence one way or the other. I'm taking a look at several containers-related issues today so I might be able to dig deeper into this one later.

**Reply by LanceMcCarthy** [reply]
@baronfel Chet, unfortunately, I was not able to get it to build the container discretely either.

To make sure I was able to see all parts of the operation, I broke up each step:

```powershell
✅
dotnet nuget update source "TelerikServer" -s "https://nuget.telerik.com/v3/index.json" -u 'api-key' -p ${{secrets.TELERIK_NUGET_KEY}} --store-password-in-clear-text

✅
dotnet restore

✅
dotnet build -c Release --no-restore

❌
dotnet publish 'DevCraftAspire.ApiService/DevCraftAspire.ApiService.csproj' /t:PublishContainer --no-build --no-restore
```

Error
```
error CONTAINER1015: Unable to access the repository 'lancemccarthy/aspirebase' at tag '8.0' in the registry 'ghcr.io'. Please confirm that this name and tag are present in the registry.
```

See the "Isolated Build Test" step here https://github.com/LanceMcCarthy/DevCraftAspire/actions/runs/9004749583/job/24738460781

**Reply by baronfel** [reply]
Thanks for confirming that, Lance! I'll dig in and do some debugging today and see if I can root cause it, then report back.

**Reply by LanceMcCarthy** [reply]
So, I've started really trying different parameter switches locally before trying this on GitHub again, and I can't get it working locally with just `dotnet build` yet

![image](https://github.com/dotnet/aspire/assets/3520532/3b9789ac-3389-4fa2-8381-f0107b9ec2aa)

Can you explain where I can learn more about all the build parameters available to me for setting base image values? I know the docs for the other params that create a new image... but not for the base image. 

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <ContainerBaseImage>ghcr.io/lancemccarthy/aspirebase:8.0</ContainerBaseImage>

    <ContainerRegistry>ghcr.io</ContainerRegistry>

    <!-- Won't this accidentally overwrite my image with the result of the build? -->
    <ContainerRepository>lancemccarthy/aspirebase</ContainerRepository>
    ...
  </PropertyGroup>

</Project>
```

**Reply by baronfel** [reply]
We have [docs for setting the base image](https://learn.microsoft.com/dotnet/core/docker/publish-as-container?pivots=dotnet-8-0#containerbaseimage), but there are really only two that matter:
* Use `ContainerFamily` if you want to keep the Microsoft-provided .NET base images, but change the _kind_ of image you use - e.g. for AOT, or to opt into chiseled base images
* Use `ContainerBaseImage` if you need anything else

There are no other knobs or adjustments that users can make to change the way the tooling finds the base image to use.

**Reply by baronfel** [reply]
@LanceMcCarthy I made a new, specific sample that illustrates the flow and for me it's working - please take a look at https://github.com/baronfel/sdk-container-demo/tree/main/src/custom-base-image and see if that's aligned with your expectations here.

I wonder if the token you're using has the correct set of permissions?

**Reply by LanceMcCarthy** [reply]
Thanks for the PoC project, I needed to see a working example of my setup to make sure I'm not going insane. Mine still isn't working after following yours, but at least I have a couple things to narrow down on. 

There seems to be a couple different factors to unpack here...

### Multi-Arch Base Image

I notice my base image targets multiple platforms ` linux/arm64,linux/amd64`, could this affect it being a base image later? The error can be interpreted in multiple ways:

> "Unable to access the repository 'lancemccarthy/aspirebase' at tag '8.0' in the registry 'ghcr.io'. _Please confirm that this name and tag are present in the registry._"

Here's my build and push:

```
    - name: Build Custom Image and push to GitHub packages
      uses: docker/build-push-action@v5
      with:
        context: '.dockerbuilds/baseimage'
        platforms: linux/arm64,linux/amd64
        push: true
        tags: |
          "${{env.BASE_IMAGE}}"
```

I will try your simple `docker build` command instead of the `docker/build-push-action@v5` for the base image build.

### Permissions Scope

You're correct that I would want to change the `packages` permission from `read` to `write`... but only if I want to _push_ an image to ghcr.io (the previous 'build custom base image' job does this).

Could it be that `azd deploy` passes custom args to `dotnet build` to set the container's parameters so it works in Azure without using ghcr.io or docker.io?

This would explain some of the behavior and errors... it is trying to push an image somewhere that I did not set because I am expecting `azd deploy` to handle the Azure side of things. If true, then you must always set publishing parameters if you want a custom base image?

Still... this doesn't explain why it fails when I run it locally (and I know the PAT has package publishing permissions).

### Next Steps

I am going to try your base image build process exactly and eliminate if that's the cause of the problem.

**Reply by LanceMcCarthy** [reply]
@baronfel AAAAAAnd that was it... once I do not set a target architecture for the base image, it is working for both `dotnet build` AND `azd deploy`.

OMG, that was so much headache, lol.

**Reply by baronfel** [reply]
Great to hear! I guess that means that the token permissions were right as rain, then :)

**Reply by LanceMcCarthy** [reply]
Hi guys, when deployed, I'm not seeing the correct base image being used in production (trace log proves the apt dependency is not there).

Can I double check my assumptions that I am setting the value correctly?

Most importantly, do I use the full path (registry and tag?

` <ContainerBaseImage>ghcr.io/lancemccarthy/aspirebase:8.0</ContainerBaseImage>`

or does it requires the other proj container properties to complete it?

**Reply by baronfel** [reply]
If you set that we use it, verbatim. There's no other modification of the base image that occurs one you take full control.

How do you tag your images when they are deployed? If you re-used a tag and your container host wasn't set to re-pull tags that could explain why the old (incorrect) base image was used?

**Reply by LanceMcCarthy** [reply]
@baronfel thanks for confirming that image is used like I expected it to.

To answer your question, here's the exact process:

### Job 1
he new base image is always created just prior to any new release (in the 1st job of the workflow):

https://github.com/LanceMcCarthy/DevCraftAspire/blob/4167d1c08b4edb0fab746459d82a3a83a8a963a7/.github/workflows/deploy.yml#L34-L54  (here's [the Dockerfile](https://github.com/LanceMcCarthy/DevCraftAspire/blob/main/.dockerbuilds/baseimage/Dockerfile) that creates the image)

### Job 2
Once that job is successful, then the .NET Aspire app is built and deploy in the 2nd job. Here's [the actual dotnet build](https://github.com/LanceMcCarthy/DevCraftAspire/blob/4167d1c08b4edb0fab746459d82a3a83a8a963a7/.github/workflows/deploy.yml#L76C33-L76C33) command.

In short, that resolves to

```
dotnet publish 'DevCraftAspire.ApiService/DevCraftAspire.ApiService.csproj' /t:PublishContainer \
  -p ContainerBaseImage="ghcr.io/lancemccarthy/aspirebase:8.0" \
  -p ContainerRepository="lancemccarthy/aspireapiservice" \
  -p ContainerImageTag="8.0" \
  -p ContainerRegistry="ghcr.io"
```

**Reply by baronfel** [reply]
I'm pretty confident the new `lancemccarthy/aspireapiservice/8.0` image is getting created and pushed to your GitHub Package Registry - the part I'm questioning is the container host that _runs_ the image. Since you're reusing the "8.0" tag, when you run `docker run ghcr.io/lancemccarthy/aspireapiservice/8.0` if the container runtime thinks it already has the `8.0` tag locally it may not try to pull a new image. You often have to tell container runtimes to explicitly pull new versions of already-present base images. For Docker this looks like `docker run --pull=always ghcr.io/lancemccarthy/aspireapiservice/8.0`. For kubernetes there's an ImagePullPolicy I believe.

**Reply by LanceMcCarthy** [reply]
This behavior would explain everything I'm seeing. I'm not sure how to set this always-pull behavior in Azure though?

I can also try using a timestamp based tag for the image, ex 2024.0522.1 (this is how we version our devtools).

**Reply by LanceMcCarthy** [reply]
Okay @baronfel, some good and bad news.

To ensure I stay sane, I use `${{github.run_number}}  for both the `aspirebase` and `aspireapiservice` containers' tag 😅

### Base build

```bash
# I use the run number to tag the base image
fullimagename="ghcr.io/lancemccarthy/aspirebase:${{github.run_number}}"

docker build -t $fullimagename -f Dockerfile .
docker push $fullimagename
```

This worked as expected:

![image](https://github.com/dotnet/aspire/assets/3520532/1bb14fd5-ba4c-4283-a439-7ed44b1e6d58)


### .NET Build

Now I can use that same number for the Aspire build 

```bash
    dotnet publish $projectPath /t:PublishContainer \

      # This is where I use the exact name, with tag, from the previous job that published the image
      -p ContainerBaseImage="${{needs.update_base_image.outputs.BASE_IMAGE_NAME}}" \
      
      # To help with troubleshooting, I am also using the same number for the apiservice container
      -p ContainerImageTag="${{github.run_number}}" \

      -p ContainerRepository="lancemccarthy/aspireapiservice" \
      -p ContainerRegistry="ghcr.io"
```

The build output looks good:

![image](https://github.com/dotnet/aspire/assets/3520532/45c61c8f-322f-4bb7-b5e2-8d417ae7635b)


### Verification

I went to the consoel on azure and checked if the neccessary package is present, but it is not... this means it is stil not using the correct base image?

![image](https://github.com/dotnet/aspire/assets/3520532/d82bce7a-da43-4f1d-bb34-009fe25be265)

Note, this is for sure a brand new base image, with a unique tag,... and I can confirm that the packages are being installed because I can see the image being built:

![image](https://github.com/dotnet/aspire/assets/3520532/cf93eb7a-8bbc-402b-958c-85d68f5299ba)

**Reply by LanceMcCarthy** [reply]
Okay, @timheuer @baronfel  I've confirmed this is either an Azure Container Registry problem or more likely, an `azd deploy` problem. It was not updating the `apiservice` container... only the `webfrontend` one.

Is there a flag I can use with azd deploy that will force all the containers to be updated? I'm thinking this will be it:

![image](https://github.com/dotnet/aspire/assets/3520532/43b3710e-3bcd-48d1-9c32-79075d99673c)


### Workaround

I went into my Aspire settings in Azure, looked at the latest revision for the apiservice container and manually updated it to use the exact image from GitHub (instead of from ACR). 

![image](https://github.com/dotnet/aspire/assets/3520532/2ea021c7-e16d-4241-b5a1-dacbfc985ad5)

### Evidence

and TADA... it is working => https://webfrontend.purpledune-16458cf6.eastus2.azurecontainerapps.io/

**Reply by baronfel** [reply]
Thanks for the detective work here! I've asked internally about the expected behavior here as well, but I'm not quite as familiar with the specifics of `az deploy`, ACA->ACR configuration, and so on. If I learn more from the internal conversations I'll make sure and report back!

**Reply by LanceMcCarthy** [reply]
Knowing what I know now, I think this happened because `azd deploy` only updated the `webfrontend` container defined in the aspire-manifest.

I am about to do another run now that uses `azd deploy -all` to verify my suspicion (it should push updates to all the containers in the manifest).

If this is true, we should definitely add something to the .NET Aspire docs on deployment. I will report back here once it's all deployed.

**Reply by LanceMcCarthy** [reply]
Okay, I have the result... unfortunately, it didn't work, @baronfel we need some input from the ACR folks.

I just ran a full deploy and all of the containers got pushed:

![image](https://github.com/dotnet/aspire/assets/3520532/9e2e2d20-7b41-4750-8867-0a8c893105b1)

I confirmed, in Azure, that my apiservice app was updated by `azd deploy` to use ACR:

![image](https://github.com/dotnet/aspire/assets/3520532/693a16b4-0ca1-44e3-9e48-d996263c325b)

but when I run it, that image from ACR is bad:

![image](https://github.com/dotnet/aspire/assets/3520532/ba2c1c4e-4b24-41de-8d7d-4af8f7168929)

I don't know why the ACR based image is incorrect, so that's where we currently stand.

Since I need to stop for the day, I've re-applied my workaround so the app is running. If the ACR folks want an actual container to review, here are the details of the most recent broken one:

```
registry: acrjmotg62yke5uc.azurecr.io
image: azd-deploy-apiservice-1716411649
tag: 8.0
```

**Reply by LanceMcCarthy** [reply]
@baronfel Now that we know where the issue lies, I have started a new discussion to solely discuss the ACR thing https://github.com/dotnet/aspire/discussions/4266

Can you please reply in the dedicated conversation when/if you hear back from the ACR folks? Thanks!

**Comment by raffaeler** [comment]
A while ago I opened [this issue](https://github.com/dotnet/aspire/issues/1852) to support dockerfiles in Aspire instead of just relying on images.

IMO this would simplify this scenario as well. Creating an intermediate image is good when you consolidate, but certainly not in the developer loop which is where Aspire should provide the stronger support.

**Reply by davidfowl** [reply]
Yep that’s on our list of things to do post GA. It assumes you’ve written a docker file though…

**Reply by raffaeler** [reply]
Fair enough :)
